core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala

```diff
@@ -62,4 +88,31 @@ private[spark] trait ShuffleManager {
 
   /** Shut down this ShuffleManager. */
   def stop(): Unit
+
+  private[this] val shuffleToAttempts = new ConcurrentHashMap[Int, ConcurrentHashMap[Int, Int]]()
+
+  /**
+   * Register a stage attempt for the given shuffle, so we can clean up all attempts when
+   * the shuffle is unregistered
+   */
+  protected def addShuffleAttempt(shuffleId: Int, stageAttemptId: Int): Unit = {
+    shuffleToAttempts.putIfAbsent(shuffleId, new ConcurrentHashMap[Int, Int]())
+    shuffleToAttempts.get(shuffleId).putIfAbsent(stageAttemptId, stageAttemptId)
```

Comment:
good point, `CopyOnWriteArraySet` makes more sense.

this actually might get called concurrently for the same shuffle -- the shuffle manager is shared by all threads of an executor, and each shuffle map task will call this method.  (It is rare that this will get called concurrently with different `stageAttemptId`s.)  `CopyOnWriteArraySet` is a good fit, though.